File: android/features/dashboard/src/main/java/com/bolsaterminal/features/dashboard/DashboardViewModel.kt
package com.bolsaterminal.features.dashboard

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bolsaterminal.core.common.UiState
import com.bolsaterminal.core.model.MarketIndex
import com.bolsaterminal.core.model.ScreenerItem
import com.bolsaterminal.domain.usecase.GetMarketOverviewUseCase
import com.bolsaterminal.domain.usecase.GetScreenerUseCase
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class DashboardUiModel(
    val indices: List<MarketIndex>,
    val topMovers: List<ScreenerItem>,
)

@HiltViewModel
class DashboardViewModel @Inject constructor(
    private val getMarketOverview: GetMarketOverviewUseCase,
    private val getScreener: GetScreenerUseCase,
) : ViewModel() {

    private val _state = MutableStateFlow<UiState<DashboardUiModel>>(UiState.Idle)
    val state: StateFlow<UiState<DashboardUiModel>> = _state.asStateFlow()
    private var consecutiveRefreshFailures = 0

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            _state.value = UiState.Loading
            try {
                _state.value = UiState.Loaded(fetch())
            } catch (e: Exception) {
                _state.value = UiState.Error(e.message.orEmpty())
            }
        }
    }

    /**
     * Periodic silent refresh — keeps showing the last good data on
     * failure, no loading spinner. After 3 consecutive failures falls back
     * to a full [load] so a sustained outage surfaces as an actionable
     * error instead of staying stuck on stale/empty data forever.
     */
    fun refresh() {
        viewModelScope.launch {
            try {
                _state.value = UiState.Loaded(fetch())
                consecutiveRefreshFailures = 0
            } catch (_: Exception) {
                consecutiveRefreshFailures++
                if (consecutiveRefreshFailures >= 3) {
                    consecutiveRefreshFailures = 0
                    load()
                }
            }
        }
    }

    private suspend fun fetch(): DashboardUiModel {
        val indices = getMarketOverview()
        val screener = getScreener()
        val topMovers = screener.sortedByDescending { kotlin.math.abs(it.changePercent) }.take(10)

        // A backend hiccup can return a valid-but-empty list rather than
        // actually failing — don't let that silently wipe good data on screen.
        val current = (_state.value as? UiState.Loaded<DashboardUiModel>)?.data
        if (indices.isEmpty() && current?.indices?.isNotEmpty() == true) {
            throw IllegalStateException("Empty market overview response")
        }
        if (topMovers.isEmpty() && current?.topMovers?.isNotEmpty() == true) {
            throw IllegalStateException("Empty screener response")
        }
        return DashboardUiModel(indices, topMovers)
    }
}

File: macos/BolsaTerminal/Features/Dashboard/DashboardViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class DashboardViewModel {
    enum State {
        case idle, loading, loaded, error(String)
    }

    var state: State = .idle
    var indices: [MarketIndex] = []
    var topMovers: [ScreenerItem] = []
    private var consecutiveRefreshFailures = 0

    func load() async {
        state = .loading
        do {
            try await fetch()
            state = .loaded
        } catch {
            state = .error((error as? LocalizedError)?.errorDescription ?? error.localizedDescription)
        }
    }

    /// Periodic background refresh — updates data without flipping to the
    /// loading state (which would blank an already-loaded screen), and
    /// silently ignores transient failures so the last good data stays on
    /// screen until the next tick succeeds. After 3 consecutive failures
    /// (sustained outage, not a blip) it falls back to a full `load()` so a
    /// real, lasting problem surfaces as an actionable error instead of the
    /// screen silently sitting on stale data forever.
    func refresh() async {
        do {
            try await fetch()
            state = .loaded
            consecutiveRefreshFailures = 0
        } catch {
            consecutiveRefreshFailures += 1
            if consecutiveRefreshFailures >= 3 {
                consecutiveRefreshFailures = 0
                await load()
            }
        }
    }

    private func fetch() async throws {
        async let indicesTask: [MarketIndex] = APIClient.shared.request(Endpoints.marketOverview)
        async let screenerTask: [ScreenerItem] = APIClient.shared.request(Endpoints.screener())
        let (idx, screener) = try await (indicesTask, screenerTask)

        // A backend hiccup can return a valid-but-empty array (HTTP 200, no
        // thrown error) rather than actually failing — treat that as a
        // failure too, so it doesn't silently wipe out good data already on
        // screen with nothing to show for it.
        guard (!idx.isEmpty || indices.isEmpty), (!screener.isEmpty || topMovers.isEmpty) else {
            throw APIError.server(status: 200, message: "Empty response")
        }

        indices = idx
        topMovers = Array(
            screener.sorted { abs($0.changePercent) > abs($1.changePercent) }.prefix(10)
        )
    }
}
